// @ionic/app-scripts/dist/deep-linking/util.spec.js
"use strict";
var path_1 = require('path');
var util = require('./util');
describe('util', function () {
    describe('extractDeepLinkPathData', function () {
        it('should return the parsed deep link metadata', function () {
            var fileContent = "\nimport { NgModule } from '@angular/core';\nimport { IonicApp, IonicModule } from 'ionic-angular';\nimport { MyApp } from './app.component';\nimport { HomePage } from '../pages/home/home';\n\nimport * as Constants from '../util/constants';\n\n@NgModule({\n  declarations: [\n    MyApp,\n    HomePage\n  ],\n  imports: [\n    getSharedIonicModule()\n  ],\n  bootstrap: [IonicApp],\n  entryComponents: [\n    MyApp,\n    HomePage\n  ],\n  providers: []\n})\nexport class AppModule {}\n\nexport function getSharedIonicModule() {\n  return IonicModule.forRoot(MyApp, {}, {\n    links: [\n      { loadChildren: '../pages/home/home.module#HomePageModule', name: 'Home' },\n      { name: \"PageOne\", loadChildren: \"../pages/page-one/page-one.module#PageOneModule\" },\n      { loadChildren: `../pages/page-two/page-two.module#PageTwoModule`, name: `PageTwo` },\n      { Component: MyComponent, name: 'SomePage'},\n      { name: 'SomePage2', Component: MyComponent2 }\n    ]\n  });\n}\n      ";
            var results = util.extractDeepLinkPathData(fileContent);
            expect(results).toBeTruthy();
            expect(results[0].component).toEqual(null);
            expect(results[0].name).toBe('Home');
            expect(results[0].modulePath).toBe('../pages/home/home.module');
            expect(results[0].namedExport).toBe('HomePageModule');
            expect(results[1].component).toEqual(null);
            expect(results[1].name).toBe('PageOne');
            expect(results[1].modulePath).toBe('../pages/page-one/page-one.module');
            expect(results[1].namedExport).toBe('PageOneModule');
            expect(results[2].component).toEqual(null);
            expect(results[2].name).toBe('PageTwo');
            expect(results[2].modulePath).toBe('../pages/page-two/page-two.module');
            expect(results[2].namedExport).toBe('PageTwoModule');
            expect(results[3].component).toEqual('MyComponent');
            expect(results[3].name).toBe('SomePage');
            expect(results[3].modulePath).toBe(null);
            expect(results[3].namedExport).toBe(null);
            expect(results[4].component).toEqual('MyComponent2');
            expect(results[4].name).toBe('SomePage2');
            expect(results[4].modulePath).toBe(null);
            expect(results[4].namedExport).toBe(null);
        });
        it('should throw an exception when there is an invalid deep link config', function () {
            // arrange
            var fileContent = "\nimport { NgModule } from '@angular/core';\nimport { IonicApp, IonicModule } from 'ionic-angular';\nimport { MyApp } from './app.component';\nimport { HomePage } from '../pages/home/home';\n\nimport * as Constants from '../util/constants';\n\n@NgModule({\n  declarations: [\n    MyApp,\n    HomePage\n  ],\n  imports: [\n    getSharedIonicModule()\n  ],\n  bootstrap: [IonicApp],\n  entryComponents: [\n    MyApp,\n    HomePage\n  ],\n  providers: []\n})\nexport class AppModule {}\n\nexport function getSharedIonicModule() {\n  return IonicModule.forRoot(MyApp, {}, {\n    links: [\n      { loadChildren: '../pages/home/home.module#HomePageModule'},\n      { name: \"PageOne\", loadChildren: \"../pages/page-one/page-one.module#PageOneModule\" },\n      { loadChildren: `../pages/page-two/page-two.module#PageTwoModule`, name: `PageTwo` },\n      { Component: MyComponent, name: 'SomePage'},\n      { name: 'SomePage2', Component: MyComponent2 }\n    ]\n  });\n}\n      ";
            // act
            var knownMessage = 'Should never get here';
            try {
                util.extractDeepLinkPathData(fileContent);
                throw new Error(knownMessage);
            }
            catch (ex) {
                // assert
                expect(ex.message).not.toEqual(knownMessage);
            }
        });
    });
    describe('getDeepLinkData', function () {
        it('should return an empty list when no valid deep links are found', function () {
            var fileContent = "\nimport { NgModule } from '@angular/core';\nimport { IonicApp, IonicModule } from 'ionic-angular';\nimport { MyApp } from './app.component';\nimport { HomePage } from '../pages/home/home';\n\nimport * as Constants from '../util/constants';\n\n@NgModule({\n  declarations: [\n    MyApp,\n    HomePage\n  ],\n  imports: [\n    getSharedIonicModule()\n  ],\n  bootstrap: [IonicApp],\n  entryComponents: [\n    MyApp,\n    HomePage\n  ],\n  providers: []\n})\nexport class AppModule {}\n\nexport function getSharedIonicModule() {\n  return IonicModule.forRoot(MyApp, {});\n}\n      ";
            var srcDir = '/Users/dan/Dev/myApp/src';
            var result = util.getDeepLinkData(path_1.join(srcDir, 'app/app.module.ts'), fileContent, false);
            expect(result).toBeTruthy();
            expect(result.length).toEqual(0);
        });
        it('should return a hydrated deep link config', function () {
            var fileContent = "\nimport { NgModule } from '@angular/core';\nimport { IonicApp, IonicModule } from 'ionic-angular';\nimport { MyApp } from './app.component';\nimport { HomePage } from '../pages/home/home';\n\nimport * as Constants from '../util/constants';\n\n@NgModule({\n  declarations: [\n    MyApp,\n    HomePage\n  ],\n  imports: [\n    getSharedIonicModule()\n  ],\n  bootstrap: [IonicApp],\n  entryComponents: [\n    MyApp,\n    HomePage\n  ],\n  providers: []\n})\nexport class AppModule {}\n\nexport function getSharedIonicModule() {\n  return IonicModule.forRoot(MyApp, {}, {\n    links: [\n      { loadChildren: '../pages/home/home.module#HomePageModule', name: 'Home' },\n      { name: \"PageOne\", loadChildren: \"../pages/page-one/page-one.module#PageOneModule\" },\n      { loadChildren: `../pages/page-two/page-two.module#PageTwoModule`, name: `PageTwo` },\n      { Component: MyComponent, name: 'SomePage'},\n    ]\n  });\n}\n      ";
            var srcDir = '/Users/dan/Dev/myApp/src';
            var result = util.getDeepLinkData(path_1.join(srcDir, 'app/app.module.ts'), fileContent, false);
            expect(result[0].modulePath).toEqual('../pages/home/home.module');
            expect(result[0].namedExport).toEqual('HomePageModule');
            expect(result[0].name).toEqual('Home');
            expect(result[0].component).toEqual(null);
            expect(result[0].absolutePath).toEqual('/Users/dan/Dev/myApp/src/pages/home/home.module.ts');
            expect(result[1].modulePath).toEqual('../pages/page-one/page-one.module');
            expect(result[1].namedExport).toEqual('PageOneModule');
            expect(result[1].name).toEqual('PageOne');
            expect(result[1].component).toEqual(null);
            expect(result[1].absolutePath).toEqual('/Users/dan/Dev/myApp/src/pages/page-one/page-one.module.ts');
            expect(result[2].modulePath).toEqual('../pages/page-two/page-two.module');
            expect(result[2].namedExport).toEqual('PageTwoModule');
            expect(result[2].name).toEqual('PageTwo');
            expect(result[2].component).toEqual(null);
            expect(result[2].absolutePath).toEqual('/Users/dan/Dev/myApp/src/pages/page-two/page-two.module.ts');
            expect(result[3].modulePath).toEqual(null);
            expect(result[3].namedExport).toEqual(null);
            expect(result[3].name).toEqual('SomePage');
            expect(result[3].component).toEqual('MyComponent');
            expect(result[3].absolutePath).toEqual(null);
        });
        it('should return a deep link data adjusted for AoT', function () {
            var fileContent = "\nimport { NgModule } from '@angular/core';\nimport { IonicApp, IonicModule } from 'ionic-angular';\nimport { MyApp } from './app.component';\nimport { HomePage } from '../pages/home/home';\n\nimport * as Constants from '../util/constants';\n\n@NgModule({\n  declarations: [\n    MyApp,\n    HomePage\n  ],\n  imports: [\n    getSharedIonicModule()\n  ],\n  bootstrap: [IonicApp],\n  entryComponents: [\n    MyApp,\n    HomePage\n  ],\n  providers: []\n})\nexport class AppModule {}\n\nexport function getSharedIonicModule() {\n  return IonicModule.forRoot(MyApp, {}, {\n    links: [\n      { loadChildren: '../pages/home/home.module#HomePageModule', name: 'Home' },\n      { name: \"PageOne\", loadChildren: \"../pages/page-one/page-one.module#PageOneModule\" },\n      { loadChildren: `../pages/page-two/page-two.module#PageTwoModule`, name: `PageTwo` },\n      { Component: MyComponent, name: 'SomePage'},\n    ]\n  });\n}\n      ";
            var srcDir = '/Users/dan/Dev/myApp/src';
            var result = util.getDeepLinkData(path_1.join(srcDir, 'app/app.module.ts'), fileContent, true);
            expect(result[0].modulePath).toEqual('../pages/home/home.module.ngfactory');
            expect(result[0].namedExport).toEqual('HomePageModuleNgFactory');
            expect(result[0].name).toEqual('Home');
            expect(result[0].component).toEqual(null);
            expect(result[0].absolutePath).toEqual('/Users/dan/Dev/myApp/src/pages/home/home.module.ngfactory.ts');
            expect(result[1].modulePath).toEqual('../pages/page-one/page-one.module.ngfactory');
            expect(result[1].namedExport).toEqual('PageOneModuleNgFactory');
            expect(result[1].name).toEqual('PageOne');
            expect(result[1].component).toEqual(null);
            expect(result[1].absolutePath).toEqual('/Users/dan/Dev/myApp/src/pages/page-one/page-one.module.ngfactory.ts');
            expect(result[2].modulePath).toEqual('../pages/page-two/page-two.module.ngfactory');
            expect(result[2].namedExport).toEqual('PageTwoModuleNgFactory');
            expect(result[2].name).toEqual('PageTwo');
            expect(result[2].component).toEqual(null);
            expect(result[2].absolutePath).toEqual('/Users/dan/Dev/myApp/src/pages/page-two/page-two.module.ngfactory.ts');
            expect(result[3].modulePath).toEqual(null);
            expect(result[3].namedExport).toEqual(null);
            expect(result[3].name).toEqual('SomePage');
            expect(result[3].component).toEqual('MyComponent');
            expect(result[3].absolutePath).toEqual(null);
        });
    });
    describe('validateDeepLinks', function () {
        it('should return false when one entry is missing name', function () {
            // arrange
            var invalidDeepLinkConfig = {
                name: null,
                component: {}
            };
            // act
            var result = util.validateDeepLinks([invalidDeepLinkConfig]);
            // assert
            expect(result).toEqual(false);
        });
        it('should return false when one entry has empty name', function () {
            // arrange
            var invalidDeepLinkConfig = {
                name: '',
                component: {}
            };
            // act
            var result = util.validateDeepLinks([invalidDeepLinkConfig]);
            // assert
            expect(result).toEqual(false);
        });
        it('should return false when missing component and (modulePath or namedExport)', function () {
            // arrange
            var invalidDeepLinkConfig = {
                name: 'someName',
                component: null,
                modulePath: null
            };
            // act
            var result = util.validateDeepLinks([invalidDeepLinkConfig]);
            // assert
            expect(result).toEqual(false);
        });
        it('should return false when missing component and (modulePath or namedExport)', function () {
            // arrange
            var invalidDeepLinkConfig = {
                name: 'someName',
                component: '',
                modulePath: ''
            };
            // act
            var result = util.validateDeepLinks([invalidDeepLinkConfig]);
            // assert
            expect(result).toEqual(false);
        });
        it('should return false when missing component and has valid modulePath but missing namedExport', function () {
            // arrange
            var invalidDeepLinkConfig = {
                name: 'someName',
                component: '',
                modulePath: 'somePath',
                namedExport: ''
            };
            // act
            var result = util.validateDeepLinks([invalidDeepLinkConfig]);
            // assert
            expect(result).toEqual(false);
        });
        it('should return true when it has a valid modulePath and namedExport', function () {
            // arrange
            var invalidDeepLinkConfig = {
                name: 'someName',
                component: '',
                modulePath: 'somePath',
                namedExport: 'someNamedExport'
            };
            // act
            var result = util.validateDeepLinks([invalidDeepLinkConfig]);
            // assert
            expect(result).toEqual(true);
        });
        it('should return true when it has a valid component', function () {
            // arrange
            var invalidDeepLinkConfig = {
                name: 'someName',
                component: 'MyComponent',
                modulePath: null,
                namedExport: null
            };
            // act
            var result = util.validateDeepLinks([invalidDeepLinkConfig]);
            // assert
            expect(result).toEqual(true);
        });
    });
});
